File: p0/p75/SortColors.java
package leetcode.p0.p75;

import java.util.Arrays;

public class SortColors {
    /**
     * Bucket Sort
     * 2 pass alg
     */
    public static void sortColors1(int[] nums) {
        int red = 0, white = 0, blue = 0;
        for (int num : nums) {
            switch (num) {
                case 0:
                    red++;
                    break;
                case 1:
                    white++;
                    break;
                case 2:
                    blue++;
            }
        }
        for (int i = 0; i < red; i++) {
            nums[i] = 0;
        }
        for (int i = red; i < red + white; i++) {
            nums[i] = 1;
        }
        for (int i = red + white; i < nums.length; i++) {
            nums[i] = 2;
        }
    }


    /**
     * Bucket Sort
     * one-pass algorithm using only constant extra space
     * From solution
     * Time complexity: O(n)
     * Space complexity: O(1)
     */
    public static void sortColors(int[] nums) {
        int p0 = 0;
        int p2 = nums.length - 1;
        int temp;
        int curr = 0;
        while (curr <= p2) {
            if (nums[curr] == 0) {
                temp = nums[curr];
                nums[curr] = nums[p0];
                nums[p0] = temp;
                p0++;
                curr++;
            } else if (nums[curr] == 2) {
                temp = nums[curr];
                nums[curr] = nums[p2];
                nums[p2] = temp;
                p2--;
            } else {
                curr++;
            }
        }
    }

    public static void main(String[] args) {
        int[] nums;
        int[] output;

        nums = new int[]{2, 0, 2, 1, 1, 0};
        output = new int[]{0, 0, 1, 1, 2, 2};

        sortColors(nums);
        System.out.println(Arrays.equals(nums, output));


        nums = new int[]{2, 0, 1};
        output = new int[]{0, 1, 2};

        sortColors(nums);
        System.out.println(Arrays.equals(nums, output));
    }
}

File: p0/p75/sort_colors.py
from typing import List


class Solution:
    """
    May 17, 2025 15:42
    Time Complexity: O(n)
    Space Complexity: O(1)
    """

    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        left = 0
        while left < n and nums[left] == 0:
            left += 1
        right = n - 1
        while right >= 0 and nums[right] == 2:
            right -= 1

        i = left
        while i <= right:
            if nums[i] == 2:
                nums[right], nums[i] = nums[i], nums[right]
                while nums[right] == 2:
                    right -= 1
            if nums[i] == 0:
                nums[left], nums[i] = nums[i], nums[left]
                left += 1

            i += 1


if __name__ == '__main__':
    sol = Solution()
    nums = [2, 0, 2, 1, 1, 0]
    sol.sortColors(nums)
    print(nums)
